### PR TITLE
Trojan fallback: support alpn address

### PR DESCRIPTION
**Pull request by zakuwaki**

### CODE REVIEW

**Comment by nekohasekai** [comment]
Why not get alpn in fallbackConnection?

File: adapter/inbound.go [hunk 1]
```diff
@@ -5,7 +5,7 @@ import (
 	"net/netip"
 
 	"github.com/sagernet/sing-box/common/process"
-	"github.com/sagernet/sing-dns"
+	dns "github.com/sagernet/sing-dns"
```
Comment (nekohasekai) [inline adapter/inbound.go:8]: Why?

File: adapter/inbound.go [hunk 2]
```diff
@@ -26,6 +26,7 @@ type InboundContext struct {
 	Protocol    string
 	User        string
 	Outbound    string
+	Alpn        string
```
Comment (nekohasekai) [inline adapter/inbound.go:29]: I don't think it's necessary

File: inbound/trojan.go [hunk 3]
```diff
@@ -45,10 +46,14 @@ func NewTrojan(ctx context.Context, router adapter.Router, logger log.ContextLog
...
-		inbound.fallbackAddr = options.Fallback.Build()
-		fallbackHandler = adapter.NewUpstreamContextHandler(inbound.fallbackConnection, nil, nil)
+	fallbackAddrs := make(map[string]M.Socksaddr)
+	for _, fallback := range options.Fallbacks {
+		if fallback.Server != "" {
+			fallbackAddrs[strings.ToLower(fallback.Alpn)] = fallback.Build()
+		}
 	}
+	inbound.fallbackAddrs = fallbackAddrs
+	fallbackHandler = adapter.NewUpstreamContextHandler(inbound.fallbackConnection, nil, nil)
```
Comment (nekohasekai) [inline inbound/trojan.go:56]: Why is fallback always enabled? you changed its behavior.

**Comment by zakuwaki** [comment]
Yes I tried, but the conn is over wrapped and alpn cannot be fetched

**Comment by nekohasekai** [comment]
Also, always rebase instead of merging new changes.

**Comment by nekohasekai** [comment]
> 是的，我试过了，但是 conn 被过度包装，无法获取 alpn

Use common.Cast

**Comment by nekohasekai** [comment]
> Why not get alpn in fallbackConnection?

Use common.Cast

**Comment by zakuwaki** [comment]
I'll fix all of this and submit a new pr. Thank you

File: option/trojan.go [hunk 4]
```diff
@@ -13,6 +13,11 @@ type TrojanUser struct {
 	Password string `json:"password"`
 }
 
+type TrojanFallback struct {
+	Alpn string `json:"alpn"`
```
Comment (nekohasekai) [inline option/trojan.go:17]: ALPN instead of Alpn, it's not a word (NextProto is fine too).

**Review by nekohasekai — CHANGES_REQUESTED** [review]

**Comment by nekohasekai** [comment]
> I'll fix all of this and submit a new pr. Thank you

Keep force commit here is just enough.

File: adapter/inbound.go (hunk 1 above)
Comment (zakuwaki) [reply adapter/inbound.go]: gofmt do this...

Comment (nekohasekai) [reply adapter/inbound.go]: Use `make fmt_install` and `make fmt`

**Comment by zakuwaki** [comment]
all done

File: inbound/trojan.go (hunk 3 above)
Comment (zakuwaki) [reply inbound/trojan.go:56]: should check fallbackAddrs > 0?

File: inbound/trojan.go [hunk 5]
```diff
@@ -153,8 +158,18 @@ func (h *Trojan) newConnection(ctx context.Context, conn net.Conn, metadata adap
...
-	metadata.Destination = h.fallbackAddr
+	var alpn string
+	if tlsConn, ok := common.Cast[*tls.Conn](conn); ok {
+		cs := tlsConn.ConnectionState()
+		alpn = strings.ToLower(cs.NegotiatedProtocol)
+	}
+
+	fallbackAddr, ok := h.fallbackAddrs[alpn]
+	if !ok {
+		fallbackAddr = h.fallbackAddrs[""]
```
Comment (nekohasekai) [inline inbound/trojan.go:169]: What if there is no default value?

File: inbound/trojan.go (hunk 3 above)
Comment (nekohasekai) [reply inbound/trojan.go:56]: Yes

File: inbound/trojan.go [hunk 6]
```diff
@@ -45,8 +46,14 @@ func NewTrojan(ctx context.Context, router adapter.Router, logger log.ContextLog
 		users: options.Users,
 	}
 	var fallbackHandler N.TCPConnectionHandler
-	if options.Fallback != nil && options.Fallback.Server != "" {
-		inbound.fallbackAddr = options.Fallback.Build()
+	fallbackAddrs := make(map[string]M.Socksaddr)
+	for _, fallback := range options.Fallbacks {
+		if fallback.Server != "" {
```
Comment (nekohasekai) [inline inbound/trojan.go:51]: If empty, it is better to return error.

**Review by nekohasekai — CHANGES_REQUESTED** [review]

**Comment by zakuwaki** [comment]
All done. Thanks for your patient guiding! I'm new in golang and really learn a lot.

File: inbound/trojan.go (hunk 6 above)
Comment (nekohasekai) [reply inbound/trojan.go:51]: Here, not in fallback.

**Comment by nekohasekai** [comment]
pushed by mistake :(

**Comment by nekohasekai** [comment]
It appears that you cannot be included as a contributor because you are using an invalid email.

**Pull request merged**